python/gpusim_server.py

```diff
@@ -73,16 +80,33 @@ def do_GET(self):
         self.send_error(404, 'Server unavailable.')
 
     def get_data_from_dbname(self, dbname):
-        if dbname not in sockets:
-            self.send_error(404, f'DB {dbname} not found in options: {sockets.keys()}') #noqa
-            raise ValueError('DB not found')
-        allsmiles, allids, allscores = [], [], []
-        src_smiles, return_count, similarity_cutoff = self.get_posted_data()
-        output_qba = self.get_data(dbname, src_smiles, return_count,
-                                    similarity_cutoff)
-        return_count, smiles, ids, scores = self.deserialize_results(output_qba)
-
-        return smiles, ids, scores, src_smiles
+        global search_mutex
+        search_mutex.lock()
+        try:
+            if dbname not in sockets:
+                self.send_error(404, f'DB {dbname} not found in options: {sockets.keys()}') #noqa
+                raise ValueError('DB not found')
+            allsmiles, allids, allscores = [], [], []
+            src_smiles, return_count, similarity_cutoff = self.get_posted_data()
+            request_num = random.randint(0, 2**31)
+            print("Processing request {}".format(request_num), file=sys.stderr)
+            output_qba = self.get_data(dbname, src_smiles, return_count,
```

Comment:
It looks like the mutex try/finally only needs to be around get_data, no?

Reply:
Not really, coudl also be the dbname sockets check, get_ata or deserialize_results.  Really I want to make sure if anything happens we unlock the mutex, it's not blindly catching anything just always making sure the mutex is unlocked, so I think it's fine to have it at this higher level

Reply:
Can you turn the mutex locking/unlocking into a context manager then?